src/websocket/sockets.py

```diff
@@ -34,3 +36,50 @@ async def log_message(sid: str, *args: List[str]) -> None:
         *args (List[str]): The arguments of the message
     """
     print(f"[SOCKET] Message from {sid}: {args}")
+
+
+@sio.on("ping")
+async def ping(sid: str, _: dict = {}) -> None:
+    """Runs whenever a client pings the server and emits a 'pong' event in response
+
+    Args:
+        sid (str): The session ID of the client pinging the server
+    """
+    print(f"[SOCKET] Ping from {sid}")
+    await sio.emit("pong", room=sid)
+
+
+@sio.on("drone_update")
+async def drone_update(sid: str, data: dict) -> None:
+    """
+    Runs whenever the drone telemetry data is received
+    Saves the telemetry data to the database and deletes records over 5 mins old
+
+    Args:
+        sid (str): The session ID of the client sending the update
+        data (dict): The telemetry data of the drone
+    """
+
+    print(f"[SOCKET] Drone Update from {sid}: {data}")
+
+    await sync_to_async(process_drone_update)(data)
+
+
+def process_drone_update(data: dict) -> None:
+    """
+    Handles the synchronous processing of the drone update
+    Saves the telemetry data to the database and deletes records over 5 mins old
+    """
+    from drone.models import DroneTelemetry
+    from drone.serializers import DroneTelemetrySerializer
+
+    # Validate telemetry data
+    telemetry = DroneTelemetrySerializer(data=data)
+    telemetry.is_valid(raise_exception=True)
+
+    telemetry.save()
+
+    # Drop records older than 5 minutes
+    cutoff_time = (datetime.now() - timedelta(minutes=5)).timestamp()
+    old_records = DroneTelemetry.objects.filter(timestamp__lt=int(cutoff_time))
+    old_records.delete()
```

Comment:
Would this introduce race conditions if there are multiple threads of this function running? Like say if:

- Thread A gets records (2, 3, 4) 
- Thread B also gets records (2, 3, 4)
- Thread A deletes (4)
- Thread B throws and error???

I'm not sure if this will throw an error, but if it does, we should probably wrap that in a try-except. Also, would it be better to store this code into the model `save()` function instead to force it to be sync?

Reply:
I looked into it a bit and it seems like the generated SQL queries won't throw errors even if they're executed at the same time. From my understanding even though queries A and B have the same queryset, when one is executing, the records will be write locked so the other query has to wait before trying to delete them.  After the first query executes, the lock is lifted and because the records no longer exist, the second query just finds nothing to delete and returns. 

Also I can put the delete code in the model save but it would pretty tightly coupled to all db writes, not sure how concerned you are with this. If you're confident that its the way to go i'm happy to move it tho 👍

Reply:
Sounds good, was just curious